### PR TITLE
Implemented [func.not_fn]

### PR DESCRIPTION
**Pull request by K-ballo**

### CODE REVIEW

File: include/slb/functional.hpp [hunk 1]
```diff
@@ -175,6 +176,82 @@ invoke(F&& f,
...
+    return !detail::invoke(std::move(fd), std::forward<Args>(args)...);
+  }
+#endif
+};
+} // namespace detail
+
+template <typename F, typename FD = typename std::decay<F>::type>
+detail::not_fn_result<FD> not_fn(F&& f) {
+  static_assert(std::is_move_constructible<FD>::value &&
+                    std::is_convertible<FD&&, FD>::value,
```
Comment (mpark) [inline include/slb/functional.hpp:247]: Why is the `is_convertible` condition here?

**Review by mpark — COMMENTED** [review]
Reviewed implementation. Need to look at tests.

File: test/functional/not_fn.cpp [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+// gcc finds calls on const rvalues ambiguous up to version 4.8.
+#if !defined(__GNUC__) || (__GNUC__ > 4) ||                                    \
+    ((__GNUC__ == 4) && (__GNUC_MINOR__ > 8))
+  int operator!() const && noexcept { return base - 4; }
+#endif
+};
+
+// Account for P0012: "Make exception specifications be part of the type
+// system".
+static const bool p0012 = !std::is_same<void(), void() noexcept>::value;
```
Comment (mpark) [inline test/functional/not_fn.cpp]: `static constexpr`?

File: test/functional/not_fn.cpp [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+// template <class F> unspecified not_fn(F&& f);
+
+struct Boolean {
+  int ret;
+  Boolean(int val) : ret(val) {}
+
+  int operator!() const noexcept { return ret; }
+};
+
+struct BooleanRef {
```
Comment (mpark) [inline test/functional/not_fn.cpp]: Why is this called `BooleanRef`...?

File: include/slb/functional.hpp [hunk 4]
```diff
@@ -175,6 +176,82 @@ invoke(F&& f,
...
+public:
+  template <typename F>
+  not_fn_result(not_fn_tag, F&& f) : fd(std::forward<F>(f)) {}
+
+  not_fn_result(not_fn_result&&) = default;
+  not_fn_result(not_fn_result const&) = default;
+
+  template <typename... Args>
+      auto operator()(Args&&... args) &
+      noexcept(noexcept(!detail::invoke(std::declval<FD&>(),
```
Comment (mpark) [inline include/slb/functional.hpp:206]: http://eel.is/c++draft/func.not_fn doesn't seem to mention these `noexcept` conditions...
Is there maybe a relevant LWG issue regarding this?

File: include/slb/functional.hpp (hunk 1 above)
Comment (K-ballo) [reply include/slb/functional.hpp]: > `FD shall satisfy the requirements of MoveConstructible__MoveConstructible_ requires both:
```
T(rv);
T u = rv;
```

Comment (K-ballo) [reply include/slb/functional.hpp:247]: `FD` shall satisfy _MoveConstructible_, which requires both `T u = rv;` and `T(rv);`

File: test/functional/not_fn.cpp (hunk 3 above)
Comment (K-ballo) [reply test/functional/not_fn.cpp]: It used to hold a reference to an `int`, member of `C`.

Is `BooleanObj` any better?

File: include/slb/functional.hpp (hunk 4 above)
Comment (K-ballo) [reply include/slb/functional.hpp:206]: Not that I know, LWG doesn't generally do conditional `noexcept` other than move/swap. Implementations are free to add them.

File: include/slb/functional.hpp (hunk 1 above)
Comment (mpark) [reply include/slb/functional.hpp:247]: huh... and here I was thinking "`is_move_constructible` probably answers whether `T` is  `MoveConstructible`". Rookie mistake.

File: include/slb/functional.hpp (hunk 4 above)
Comment (mpark) [reply include/slb/functional.hpp:206]: > Implementations are free to add them.

👍

File: test/functional/not_fn.cpp [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+std::false_type const throws{};
+std::integral_constant<bool, p0012> const p0012_nothrows{};
+
+template <typename R, bool IsNothrow, typename F, typename... Args>
+void check_not_fn(R&& r,
+                  std::integral_constant<bool, IsNothrow>,
+                  F&& nf,
+                  Args&&... args) {
+  CHECK(std::forward<F>(nf)(std::forward<Args>(args)...) == r);
+  CHECK(std::is_same<decltype(nf(std::forward<Args>(args)...)), R>::value);
```
Comment (mpark) [inline test/functional/not_fn.cpp]: We should forward the `nf` here I think, although it doesn't seem like we have any test cases that break this currently.

**Review by mpark — APPROVED** [review]
LGTM

File: test/functional/not_fn.cpp (hunk 5 above)
Comment (K-ballo) [reply test/functional/not_fn.cpp]: We should add one then.

**Pull request merged**